README.md
# tscausal
Time Series Causal Analysis


Changelog
# 2020-12-29 v0.3.0 [AG]
- License changed to APACHE to be compliant with use of code from CausalImpact.
- Reporting code moved to separate file and function and copyright notice addded
to that file to conform to requirements for using the CausalImpact reporting code.

# 2020-12-16 v0.2.0 [AG]
- Added option to include or exclude cumulative analysis since this only makes sense for flow variables.
- Added extra plotting options to pass to the distribution plots


R/plots.R
plot.tscausal <- function(x, y = NULL, median_color = "black", median_type = 1, median_width = 2, 
                          gradient_color = "orange", interval_color = "red", interval_type = 2, 
                          interval_width = 2, ylim = NULL, ylab = "", n_original = NULL, ...)
{
  p <- process.tscausal(x, n_original = n_original)
  if (is.null(n_original)) n_original <- NROW(x$pre_actual)
  par(bg = "white", mar = c(1,2,0.5,3))
  if (x$include_cumulative) {
    layout(mat = matrix(c(1,2,3), nrow = 3))
    plot(p$prediction, interval_quantiles = c(p$alpha/2, 1 - p$alpha/2), interval_color = interval_color, gradient_color = gradient_color, x_axes = FALSE, 
         median_color = median_color, median_type = median_type, median_width = median_width, interval_type = interval_type, 
         interval_width = interval_width, ylim = ylim, ylab = ylab)
    abline(v = as.Date(p$intervention_date), col = "green", lty = 2, lwd = 2)
    mtext("Original", side = 4, adj = 0.5, padj = 0.5, cex = 0.8, font = 2, family = "mono")
    legend("topleft", c("Actual","Counterfactual"), col = c("red","black"), lty = 1, bty = "n", lwd = c(1,2))
    p$point_distribution = na.fill(p$point_distribution, fill = 0)
    class(p$point_distribution) <- "tsmodel.distribution"
    plot(p$point_distribution, interval_quantiles = c(p$alpha/2, 1 - p$alpha/2),interval_color = interval_color, gradient_color = gradient_color, x_axes = FALSE, 
         median_color = median_color, median_type = median_type, median_width = median_width, interval_type = interval_type, 
         interval_width = interval_width, ylim = ylim, ylab = ylab)
    abline(v = as.Date(p$intervention_date), col = "green", lty = 2, lwd = 2)
    mtext("Pointwise", side = 4, adj = 0.5, padj = 0.5, cex = 0.8, font = 2, family = "mono")
    par(mar = c(3,2,0.5,3))
    p$cumulative_effect_dist = na.fill(p$cumulative_effect_dist, fill = 0)
    class(p$cumulative_effect_dist) <- "tsmodel.distribution"
    plot(p$cumulative_effect_dist, interval_quantiles = c(p$alpha/2, 1 - p$alpha/2), interval_color = interval_color, gradient_color = gradient_color, x_axes = TRUE, 
         median_color = median_color, median_type = median_type, median_width = median_width, interval_type = interval_type, 
         interval_width = interval_width, ylim = ylim, ylab = ylab)
    abline(v = as.Date(p$intervention_date), col = "green", lty = 2, lwd = 2)
    mtext("Cumulative", side = 4, adj = 0.5, padj = 0.5, cex = 0.8, font = 2, family = "mono")
    par(mfrow = c(1,1), mar = c(5, 4, 4, 2) + 0.1)
  } else {
    layout(mat = matrix(c(1,2), nrow = 2))
    plot(p$prediction, interval_quantiles = c(p$alpha/2, 1 - p$alpha/2), interval_color = interval_color, gradient_color = gradient_color, x_axes = FALSE, 
         median_color = median_color, median_type = median_type, median_width = median_width, interval_type = interval_type, 
         interval_width = interval_width, ylim = ylim, ylab = ylab)
    abline(v = as.Date(p$intervention_date), col = "green", lty = 2, lwd = 2)
    mtext("Original", side = 4, adj = 0.5, padj = 0.5, cex = 0.8, font = 2, family = "mono")
    legend("topleft", c("Actual","Counterfactual"), col = c("red","black"), lty = 1, bty = "n", lwd = c(1,2))
    p$point_distribution = na.fill(p$point_distribution, fill = 0)
    class(p$point_distribution) <- "tsmodel.distribution"
    par(mar = c(3,2,0.5,3))
    plot(p$point_distribution, interval_quantiles = c(p$alpha/2, 1 - p$alpha/2), interval_color = interval_color, gradient_color = gradient_color, x_axes = TRUE, 
         median_color = median_color, median_type = median_type, median_width = median_width, interval_type = interval_type, 
         interval_width = interval_width, ylim = ylim, ylab = ylab)
    abline(v = as.Date(p$intervention_date), col = "green", lty = 2, lwd = 2)
    mtext("Pointwise", side = 4, adj = 0.5, padj = 0.5, cex = 0.8, font = 2, family = "mono")
    par(mfrow = c(1,1), mar = c(5, 4, 4, 2) + 0.1)
  }
}



process.tscausal <- function(object, n_original = NULL){
  # Check input
  alpha <- object$alpha
  prob_lower <- alpha / 2      # e.g., 0.025 when alpha = 0.05
  prob_upper <- 1 - alpha / 2  # e.g., 0.975 when alpha = 0.05
  if (!is.null(object$fitted)) {
    if (is(object$fitted,"tsmodel.distribution")) {
      if (!is.null(n_original)) {
        N <- ncol(object$fitted)
        n <- pmin(pmax(1, n_original), N)
        fx <- object$fitted[,(N - n):N]
      } else{
        fx <- object$fitted
      }
      fit_dist <- fx
    } else{
      if (!is.null(n_original)) {
        fx <- tail(object$fitted, pmax(1, n_original))
      } else {
        fx <- object$fitted
      }
      fit_dist <- matrix(as.numeric(fx), ncol = NROW(fx), nrow = NROW(object$predicted), byrow = TRUE)
    }
  } else {
    if (is.null(n_original)) n_original <- NROW(object$pre_actual)
    fit_dist <- matrix(NA, ncol = n_original, nrow = NROW(object$predicted), byrow = TRUE)
  }
  colnames(fit_dist) <- as.character(tail(index(object$pre_actual), NCOL(fit_dist)))
  response <- rbind(tail(object$pre_actual, NCOL(fit_dist)), object$post_actual)
  total_pred_dist <- cbind(fit_dist, object$predicted)
  class(total_pred_dist) <- "tsmodel.distribution"
  actual_dist <- cbind(matrix(as.numeric(object$pre_actual[colnames(fit_dist)]), ncol = ncol(fit_dist), nrow = nrow(fit_dist), byrow = TRUE),
                      matrix(as.numeric(object$post_actual), ncol = NROW(object$post_actual), nrow = nrow(fit_dist),byrow = TRUE))
  colnames(actual_dist) <- c(colnames(fit_dist), as.character(index(object$post_actual)))
  class(actual_dist) <- "tsmodel.distribution"
  intN <- which(colnames(actual_dist) == object$intervention_date)
  N <- ncol(total_pred_dist)
  point_dist <- actual_dist - total_pred_dist
  class(point_dist) <- "tsmodel.distribution"
  pdist <- point_dist
  pdist[,1:intN] <- 0
  cumulative_effect <- t(apply(pdist, 1, cumsum))
  class(cumulative_effect) <- "tsmodel.distribution"
  pout <- list(distribution = object$predicted, original_series = zoo(rbind(object$pre_actual[colnames(fit_dist)], object$post_actual)), 
               h = ncol(object$predicted))
  class(pout) <- "tsmodel.predict"
  out <- list(prediction = pout, point_distribution = point_dist, cumulative_effect_dist = cumulative_effect, 
              intervention_date = object$intervention_date, alpha = object$alpha)
  return(out)
}


R/utils.R
# Code used from CausalImpact
prettify_percentage <- function(x, round_digits = 0L)
{
  round.digits <- as.integer(round_digits)
  return(sprintf("%+0.*f%%", round_digits, 100 * x))
}

prettify_single_number <- function(x, letter, round_digits)
{
  if (is.na(x) && !is.nan(x)) {
    return("NA")
  } else if (!is.finite(x)) {
    return(as.character(x))
  } else if ((letter == "" && abs(x) >= 1e9) || letter == "B") {
    return(sprintf("%0.*fB", round_digits, x / 1e9))
  } else if ((letter == "" && abs(x) >= 1e6) || letter == "M") {
    return(sprintf("%0.*fM", round_digits, x / 1e6))
  } else if ((letter == "" && abs(x) >= 1e3) || letter == "K") {
    return(sprintf("%0.*fK", round_digits, x / 1e3))
  } else if (abs(x) >= 1 || x == 0) {
    return(sprintf("%0.*f", round_digits, x))
  } else {
    # Calculate position of first non-zero digit after the decimal point
    first.nonzero <- -floor(log10(abs(x)))
    return(sprintf("%0.*f", round_digits + first.nonzero - 1, x))
  }
}

prettify_number <- function(x, letter = "", round_digits = 1L)
{
  round_digits <- as.integer(round_digits[1])
  letter <- rep(letter, length.out = length(x))
  output <- sapply(seq_along(x), function(index) {
    prettify_single_number(x[index], letter[index], round_digits)
    })
  return(output)
}

string_trim <- function(x)
{ 
  gsub("^\\s+|\\s+$", "", x)
}

format_number <- function(x, digits)
{
  string_trim(format(x, digits = digits))
}

format_percent <- function(x, digits) 
{
  string_trim(paste0(format(x * 100, digits = digits), "%"))
}

format_confidence_intervals <- function(a, b, digits)
{
  paste0("[", string_trim(format(a, digits = min(digits, 2))),
         ", ", string_trim(format(b, digits = min(digits, 2))),
         "]")
}

format_percent_confidence_intervals <- function(a, b, digits)
{
  paste0("[", string_trim(format(a * 100, digits = min(digits, 2))),
         "%, ", string_trim(format(b * 100, digits = min(digits, 2))),
         "%]")
}

identify_number_abbreviation <- function(abbreviated_number)
{
  letter <- substr(abbreviated_number, nchar(abbreviated_number),
                   nchar(abbreviated_number))
  letter[!is.element(letter, c("B", "M", "K"))] <- "none"
  return(letter)
}


R/CausalImpact_import.R
# Copyright 2014-2020 Google Inc. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# The reporting code below was taken from CausalImpact and adapted for use in the tscausal package.

report_statement <- function(actual, prediction, object, alpha, p, sig, pos, ci_coverage, prediction_lower, prediction_upper, absolute_effect, 
                              relative_effect, relative_effect_lower, relative_effect_upper, 
                              absolute_effect_lower, absolute_effect_upper)
{
    stmt <- NULL
    stmt <- paste0(stmt, "\n\nDuring the post-intervention period, the response ",
                   "variable had an average value of approx. ", actual[1],
                   ". ", if (sig) "By contrast, in " else "In ",
                   "the absence of an intervention, ",
                   "we would have expected an average response of ",
                   prediction[1], ". The ", ci_coverage, " interval of this ",
                   "counterfactual prediction is [", prediction_lower[1],
                   ", ", prediction_upper[1], "]. Subtracting this ",
                   "prediction from the observed response yields an estimate ",
                   "of the causal effect the intervention had on the response ",
                   "variable. This effect is ", absolute_effect[1], " with a ",
                   ci_coverage, " interval of [", absolute_effect_lower[1], ", ",
                   absolute_effect_upper[1], "]. For a discussion of ",
                   "the significance of this effect, see below.")
    if (object$include_cumulative) {
        # Summarize sums
        stmt <- paste0(stmt, "\n\nSumming up the individual data points during ",
                       "the post-intervention period (which can only sometimes be ",
                       "meaningfully interpreted), the response variable had an ",
                       "overall value of ", actual[2], ". ",
                       if (sig) "By contrast, had " else "Had ",
                       "the intervention not taken place, we would have expected ",
                       "a sum of ", prediction[2], ". The ", ci_coverage, " interval of ",
                       "this prediction is [", prediction_lower[2], ", ", prediction_upper[2],
                       "].")
    }
    # Summarize relative numbers (in which case row [1] = row [2])
    stmt <- paste0(stmt, "\n\nThe above results are given in terms of ",
                   "absolute numbers. In relative terms, the response variable ",
                   "showed ", if (pos) "an increase of " else "a decrease of",
                   relative_effect[1], ". The ", ci_coverage, " interval of this ",
                   "percentage is [", relative_effect_lower[1], ", ",
                   relative_effect_upper[1], "].")
    # Comment on significance
    if (sig && pos) {
        stmt <- paste0(stmt, "\n\nThis means that the positive effect observed ",
                       "during the intervention period is statistically ",
                       "significant and unlikely to be due to random ",
                       "fluctuations. ",
                       "It should be noted, however, that the question of whether ",
                       "this increase also bears substantive significance can ",
                       "only be answered by comparing the absolute effect (",
                       absolute_effect[1], ") to the original goal of ",
                       "the underlying intervention.")
    } else if (sig && !pos) {
        stmt <- paste0(stmt, "\n\nThis means that the negative effect observed ",
                       "during the intervention period is statistically ",
                       "significant. If the experimenter had expected a positive ",
                       "effect, it is recommended to ",
                       "double-check whether anomalies in the control variables ",
                       "may have caused an overly optimistic expectation of ",
                       "what should have happened in the response variable in the ",
                       "absence of the intervention.")
    } else if (!sig && pos) {
        stmt <- paste0(stmt, "\n\nThis means that, although the intervention ",
                       "appears to have caused a positive effect, this effect ",
                       "is not statistically significant when considering the ",
                       "entire post-intervention period as a whole. Individual ",
                       "days or shorter stretches within the intervention period ",
                       "may of course still have had a significant effect, as ",
                       "indicated whenever the lower limit of the impact ",
                       "time series (lower plot) was above zero.")
    } else if (!sig && !pos) {
        stmt <- paste0(stmt, "\n\nThis means that, although it may look as ",
                       "though the intervention has exerted a negative effect ",
                       "on the response variable when considering the ",
                       "intervention period as a whole, this effect is not ",
                       "statistically significant, and so cannot be ",
                       "meaningfully interpreted.")
    }
    if (!sig) {
        stmt <- paste0(stmt, " The apparent effect could be the result of ",
                       "random fluctuations that are unrelated to the ",
                       "intervention. This is often the case when the ",
                       "intervention period is very long and includes much ",
                       "of the time when the effect has already worn off. ",
                       "It can also be the case when the intervention period ",
                       "is too short to distinguish the signal from the noise. ",
                       "Finally, failing to find a significant effect can ",
                       "happen when there are not enough control variables or ",
                       "when these variables do not correlate well with ",
                       "the response variable during the learning period.")
    }
    if (p < alpha[1]) {
        stmt <- paste0(stmt, "\n\nThe probability of obtaining this effect by ",
                       "chance is very small (Bayesian one-sided tail-area ",
                       "probability p = ", round(p, 3), "). This means the causal ",
                       "effect can be considered statistically significant.")
    } else {
        stmt <- paste0(stmt, "\n\nThe probability of obtaining this ",
                       "effect by chance is p = ", round(p, 3), ". This ",
                       "means the effect may be spurious and would generally ",
                       "not be considered statistically significant.")
    }
    return(stmt)
}

R/causal.R
tscausal.tsmodel.distribution <- function(object, actual, fitted = NULL, alpha = 0.05, include_cumulative = TRUE, ...) {
  # Check input
  pindices <- as.Date(colnames(object))
  post_actual <- actual[pindices]
  ix <- which(index(actual) %in% pindices)
  pre_actual <- actual[1:(ix[1] - 1)]
  intervention_date <- tail(index(pre_actual),1)
  if (!is.null(fitted)) pre_actual <- pre_actual[index(fitted)]
  if (NROW(post_actual) != ncol(object)) {
    stop("\ntscausal->error: length of y not equal to number of columns of forecast distribution (object).")
  }
  # We will compare the matrix of predicted trajectories (e.g., 900 x 201)
  # with a matrix of replicated observations (e.g., 900 x 201)
  y_dist <- matrix(as.numeric(post_actual), nrow = nrow(object), ncol = ncol(object), byrow = TRUE)
  point_pred <- colMeans(object)
  # Define quantiles
  alpha <- alpha[1]
  if (alpha > 1 | alpha < 0) stop("\ntscausal: alpha (coverage) must be between 0 and 1")
  prob_lower <- alpha / 2      # e.g., 0.025 when alpha = 0.05
  prob_upper <- 1 - alpha / 2  # e.g., 0.975 when alpha = 0.05
  # Compile summary statistics
  if (include_cumulative) {
    rep_n <- 2
    col_names <- c("Average", "Cumulative")
    summary_table <- data.table(type = col_names,
                                actual = c(mean(post_actual), sum(post_actual)),
                                prediction = c(mean(point_pred), sum(point_pred)),
                                prediction_lower = c(quantile(rowMeans(object), prob_lower), quantile(rowSums(object), prob_lower)),
                                prediction_upper = c(quantile(rowMeans(object), prob_upper), quantile(rowSums(object), prob_upper)),
                                prediction_sd = c(sd(rowMeans(object)),sd(rowSums(object))),
                                absolute_effect = c(mean(post_actual) - mean(point_pred), sum(post_actual) - sum(point_pred)),
                                absolute_effect_lower  = c(quantile(rowMeans(y_dist - object), prob_lower), quantile(rowSums(y_dist - object), prob_lower)),
                                absolute_effect_upper = c(quantile(rowMeans(y_dist - object), prob_upper), quantile(rowSums(y_dist - object), prob_upper)),
                                absolute_effect_sd = c(sd(rowMeans(y_dist - object)), sd(rowSums(y_dist - object))))
  } else {
    rep_n <- 1
    col_names <- "Average"
    summary_table <- data.table(type = col_names,
                                actual = mean(post_actual),
                                prediction = mean(point_pred),
                                prediction_lower = quantile(rowMeans(object), prob_lower),
                                prediction_upper = quantile(rowMeans(object), prob_upper),
                                prediction_sd = sd(rowMeans(object)),
                                absolute_effect = mean(post_actual) - mean(point_pred),
                                absolute_effect_lower  = quantile(rowMeans(y_dist - object), prob_lower),
                                absolute_effect_upper = quantile(rowMeans(y_dist - object), prob_upper),
                                absolute_effect_sd = sd(rowMeans(y_dist - object)))
  }
  summary_table[,relative_effect := absolute_effect/prediction]
  summary_table[,relative_effect_lower := absolute_effect_lower/prediction]
  summary_table[,relative_effect_upper := absolute_effect_upper/prediction]
  summary_table[,relative_effect_sd := absolute_effect_sd/prediction]
  # Add interval coverage, defined by alpha
  summary_table[,alpha := rep(alpha,rep_n)]
  # Add one-sided tail-area probability of overall impact, p
  y_samples_post_sum <- rowSums(object)
  y_post_sum <- sum(post_actual)
  p <- min(sum(c(y_samples_post_sum, y_post_sum) >= y_post_sum),
           sum(c(y_samples_post_sum, y_post_sum) <= y_post_sum))/(length(y_samples_post_sum) + 1)
  if (p > 1 | p < 0) stop("\ntscausal-->error: estimated probability outside admissible range [0,1]...check.")
  summary_table[,p :=  p]
  out = list(summary_table = summary_table, predicted = object, post_actual = post_actual, pre_actual = pre_actual,
             fitted = fitted, intervention_date = intervention_date, alpha = alpha, include_cumulative = include_cumulative)
  class(out) = "tscausal"
  return(out)
}

print.tscausal = function(x, digits = 4, ...)
{
  summary_table <- x$summary_table
  alpha <- x$alpha
  # Print title
  cat("Predictive inference {tscausal}\n")
  if (is.null(summary_table)) {
    cat("(Inference aborted)\n")
    return(invisible(NULL))
  }
  # Compile data frame with formatted numbers
  if (x$include_cumulative) {
    sep <- c("", "")
  } else {
    sep <- ""
  }
  f_summary <- data.table(
    actual = format_number(summary_table$actual, digits = digits),
    prediction = paste0(format_number(summary_table$prediction, digits = digits), " (", format_number(summary_table$prediction_sd, digits = digits), ")"),
    prediction_ci = format_confidence_intervals(summary_table$prediction_lower, summary_table$prediction_upper, digits = digits),
    Separator1 = sep,
    absolute_effect = paste0(format_number(summary_table$absolute_effect, digits = digits)," (", format_number(summary_table$absolute_effect_sd, digits = digits), ")"),
    absolute_effect_ci = format_confidence_intervals(summary_table$absolute_effect_lower, summary_table$absolute_effect_upper, digits = digits),
    Separator2 = sep,
    relative_effect = paste0(format_percent(summary_table$relative_effect, digits = digits), " (", format_percent(summary_table$relative_effect_sd, digits = digits), ")"),
    relative_effect_ci = format_percent_confidence_intervals(summary_table$relative_effect_lower, summary_table$relative_effect_upper, digits = digits))
  # Invert and format as table
  f_summary <- t(f_summary)
  if (x$include_cumulative) {
    colnames(f_summary) <- c("Average","Cumulative")
  } else {
    colnames(f_summary) <- "Average"
  }
  ci_label <- paste0(round((1 - alpha) * 100), "% CI")
  row.names(f_summary) <- c("Actual", "Prediction (s.d.)", ci_label,
                            " ",
                            "Absolute effect (s.d.)", paste(ci_label, ""),
                            "  ",
                            "Relative effect (s.d.)", paste(ci_label, " "))
  cat("\n")
  print.default(f_summary, print.gap = 3L, quote = FALSE)
  cat("\n")
  # Print overall tail-area probability
  p <- summary_table$p[1]
  cat(paste0("Predictive Distribution tail-area probability p:   ", round(p, 5), "\n"))
  cat(paste0("Predictive Distribution prob. of a causal effect:  ", round((1 - p) * 100, ifelse(p < 0.01, 5, ifelse(p < 0.05, 3, 0))), "%\n"))
  cat("\n")
}

# report statement summary output is based on CausalImpace Code
tsreport.tscausal = function(object, digits = 4, doc_template = NULL, type = c("screen", "pdf", "doc", "html"), output_dir = "/", args = list(name = "Causal Analysis", frequency = NULL, model = NULL), ...)
{
  type <- match.arg(type[1], c("screen", "pdf", "doc", "html"))
  if (type == "screen") {
    summary_table = object$summary_table
    actual <- prettify_number(summary_table$actual, round_digits = digits)
    letter <- identify_number_abbreviation(actual)
    prediction <- prettify_number(summary_table$prediction, letter, 2)
    prediction_lower <- prettify_number(summary_table$prediction_lower, letter, digits)
    prediction_upper <- prettify_number(summary_table$prediction_upper, letter, digits)
    absolute_effect <- prettify_number(summary_table$absolute_effect, letter, digits)
    absolute_effect_lower <- prettify_number(summary_table$absolute_effect_lower, letter, digits)
    absolute_effect_upper <- prettify_number(summary_table$absolute_effect_upper, letter, digits)
    relative_effect <- prettify_percentage(summary_table$relative_effect)
    relative_effect_lower <- prettify_percentage(summary_table$relative_effect_lower)
    relative_effect_upper <- prettify_percentage(summary_table$relative_effect_upper)
    # Evaluate significance and direction of the effect (increase or decrease)
    sig <- (!((summary_table$relative_effect_lower[1] < 0) && (summary_table$relative_effect_upper[1] > 0)))
    pos <- summary_table$relative_effect[1] > 0
    p <- summary_table$p[1]
    # Interval name
    ci_coverage <- paste0(round((1 - summary_table$alpha[1]) * 100), "%")
    # Initialize statement
    # Summarize averages
    stmt <- report_statement(actual, prediction, object, summary_table$alpha[1], p, sig, pos, ci_coverage, prediction_lower, prediction_upper, absolute_effect, 
                                 relative_effect, relative_effect_lower, relative_effect_upper, 
                                 absolute_effect_lower, absolute_effect_upper)
    return(stmt)
  } else {
    saveRDS(object, file = paste0(output_dir,"/causal_tmp.rds"))
    fname <- args$name
    if (is.null(fname)) fname <- paste0(fname,"_tscausal_report_",format(as.Date(Sys.Date()),"%Y%m%d")) else fname = paste0("tscausal_report_",format(as.Date(Sys.Date()),"%Y%m%d"))
    if (type == "pdf") {
      suppressWarnings(rmarkdown::render(file.path(find.package('tscausal'),'scripts/causal_report_pdf.Rmd'),
                                         output_dir = output_dir,
                                         output_file = paste0(fname,".pdf"),
                                         intermediates_dir = tempdir(),
                                         output_format = pdf_document2(toc = FALSE),
                                         params = list(dir = as.character(output_dir), name = args$name, frequency = args$frequency, model = args$model), quiet = TRUE))
    } else if (type == "doc") {
      file.copy(file.path(find.package('tscausal'),'scripts/causal_report_docx.Rmd'), file.path(paste0(output_dir,"/causal_report_docx.Rmd")))
      suppressWarnings(rmarkdown::render(file.path(paste0(output_dir,"/causal_report_docx.Rmd")),
                                         output_dir = output_dir,
                                         output_file = paste0(fname,".docx"),
                                         intermediates_dir = tempdir(),
                                         output_format = word_document2(toc = FALSE, reference_docx = doc_template, title = "Time Series Causal Analysis"),
                                         params = list(dir = as.character(output_dir), name = args$name, frequency = args$frequency, model = args$model), quiet = F))
      
    } else {
      suppressWarnings(rmarkdown::render(file.path(find.package('tscausal'),'scripts/causal_report_html.Rmd'),
                                         output_dir = output_dir,
                                         output_file = paste0(fname,".html"),
                                         intermediates_dir = tempdir(),
                                         output_format = html_document2(toc = FALSE),
                                         params = list(dir = as.character(output_dir), name = args$name, frequency = args$frequency, model = args$model), quiet = TRUE))
    }
  }
}

.latex.output = function(x, digits = 4, ...)
{
  summary_table <- x$summary_table
  actual <- prettify_number(summary_table$actual, round_digits = digits)
  letter <- identify_number_abbreviation(actual)
  prediction <- prettify_number(summary_table$prediction, letter, 2)
  prediction_lower <- prettify_number(summary_table$prediction_lower, letter, digits)
  prediction_upper <- prettify_number(summary_table$prediction_upper, letter, digits)
  absolute_effect <- prettify_number(summary_table$absolute_effect, letter, digits)
  absolute_effect_lower <- prettify_number(summary_table$absolute_effect_lower, letter, digits)
  absolute_effect_upper <- prettify_number(summary_table$absolute_effect_upper, letter, digits)
  relative_effect <- prettify_percentage(summary_table$relative_effect)
  relative_effect_lower <- prettify_percentage(summary_table$relative_effect_lower)
  relative_effect_upper <- prettify_percentage(summary_table$relative_effect_upper)
  # Evaluate significance and direction of the effect (increase or decrease)
  sig <- (!((summary_table$relative_effect_lower[1] < 0) && (summary_table$relative_effect_upper[1] > 0)))
  pos <- summary_table$relative_effect[1] > 0
  p <- summary_table$p[1]
  # Interval name
  ci_coverage <- paste0(round((1 - summary_table$alpha[1]) * 100), "%")
  stmt <- report_statement(actual, prediction, object = x, summary_table$alpha[1], p, sig, pos, ci_coverage, prediction_lower, prediction_upper, absolute_effect, 
                           relative_effect, relative_effect_lower, relative_effect_upper, 
                           absolute_effect_lower, absolute_effect_upper)
  stmt <- cat(stmt, " ")
  return(stmt)
}

.table.print.tscausal = function(x, digits = 4, ...)
{
  summary_table <- x$summary_table
  alpha <- x$alpha
  # Print title
  # Compile data frame with formatted numbers
  f_summary <- data.table(
    actual = format_number(summary_table$actual, digits = digits),
    prediction = paste0(format_number(summary_table$prediction, digits = digits), " (", format_number(summary_table$prediction_sd, digits = digits), ")"),
    prediction_ci = format_confidence_intervals(summary_table$prediction_lower, summary_table$prediction_upper, digits = digits),
    Separator1 = c("", ""),
    absolute_effect = paste0(format_number(summary_table$absolute_effect, digits = digits)," (", format_number(summary_table$absolute_effect_sd, digits = digits), ")"),
    absolute_effect_ci = format_confidence_intervals(summary_table$absolute_effect_lower, summary_table$absolute_effect_upper, digits = digits),
    Separator2 = c("", ""),
    relative_effect = paste0(format_percent(summary_table$relative_effect, digits = digits), " (", format_percent(summary_table$relative_effect_sd, digits = digits), ")"),
    relative_effect_ci = format_percent_confidence_intervals(summary_table$relative_effect_lower, summary_table$relative_effect_upper, digits = digits))
  # Invert and format as table
  f_summary <- t(f_summary)
  if (x$include_cumulative) {
    colnames(f_summary) <- c("Average", "Cumulative")
  } else {
    colnames(f_summary) <- "Average"
  }
  ci_label <- paste0(round((1 - alpha) * 100), "% CI")
  row.names(f_summary) <- c("Actual", "Prediction (s.d.)", ci_label,
                            " ",
                            "Absolute effect (s.d.)", paste(ci_label, ""),
                            "  ",
                            "Relative effect (s.d.)", paste(ci_label, " "))
  return(f_summary)
}